enhancements/split-sdk-into-multiple-projects.md

```diff
@@ -0,0 +1,221 @@
+at different repos to pick up your dep changes, these are obstacles we can
+overcome. They also aren’t that much different from other dependencies we
+work with.
+
+### Goals
+
+Split the operator-sdk repo into 4 separate repos:
+
+1. CLI and Go Operator
+1. SDK library
+1. Helm operator
+1. Ansible operator
+
+This SDK library split would be a requirement for a 1.0 release. The Ansible
+and Helm operators could go afterwards if need be.
+
+### Non-Goals
+
+* We will not solve the downstream SDK in this enhancement.
+* We will not solve all of the items required for a 1.0 release.
+* Defining when version bumps occur in each repo.
+
+## Proposal
+
+Split the operator-sdk repo into 4 separate repos as outlined in the Goals
+section above.
+
+1. CLI and Go Operator
+   * contains the Go operator code
+   * contains the CLI code
+   * related scaffolding
+   * lives in operator-framework/operator-sdk
+   * will have a deprecation policy
+     * we can deprecate in minor versions; removal becomes a major version bump
+   * this split is dependent on Kubebuilder Phase 2.
+1. SDK library
+   * contains utility code like package status, status condition, annotated
+     watcher, etc.
+   * this is something that operator projects (including Helm and Ansible
+     operators) would import
+   * depends on controller-runtime
+   * lives in operator-framework/sdk-lib (or some other name TBD)
+   * this library should probably *not* be 1.0; it's most likely to change often
+   * use Go apidiff tool to determine major version bumps
+1. Helm operator
+   * contains Helm operator code
+   * related scaffolding
+   * lives in operator-framework/helm-operator
+1. Ansible operator
+   * contains Ansible operator code
+   * related scaffolding
+   * lives in operator-framework/ansible-operator
+
+![Image of project dependencies](split-sdk-into-multiple-projects-deps.png)
+
+### Tasks
+
+1. Triage pkg directory to determine what goes to which repo
+1. Use git commands to extract directories or files with history into these
+   new repos. Exact commands will be TBD.
+1. Evaluate if making a separate repo for the hack directory is necessary or
+   if we just copy it to each of the 4 repos.
+1. CLI and Go Operator
+   * keep operator-framework/operator-sdk
+   * cmd, internal directories remain here
+   * prior to 1.0 release, review entire CLI; axe all legacy support for
+     existing project layout
+   * consider making `generate csv` separate like `controller-gen` to support
+     OLM's plan for CSVless bundles (could be done later)
+1. SDK library
+   * create new repo operator-framework/sdk-lib
+   * export the directories / files based on task 1 triage.
+1. Helm operator and plugin
+   * create new repo operator-framework/helm-operator
+   * export the pkg/helm tree to the new repo at the top level maintaining
+     history
+     * Deprecate pkg/helm and replace with [Joe's helm repo](https://github.com/joelanford/helm-operator)
```

Comment:
Can you elaborate on why/how we're replacing this package with @joelanford's repo?

Reply:
Great question. Something that probably needs more discussion.

The intent of the new repo is to:
- Be fully backwards compatible with the existing helm operator implementation
- Re-organize the underlying libraries to actually make them expose a sane public API and make unit testing easier.
- Make it extensible so that others can add their own Go code customizations for hybrid use cases.
- Cleanup a bunch of issues that make the current implementation difficult to add new features and fix bugs while ensuring that things don't break.

In addition to the feature set of the current implementation, we also need to make sure we have all the necessary tooling to continue publishing the helm operator base image upstream and downstream for all of the required architectures.

Reply:
I've updated the text that this will be covered in a separate proposal.  I also updated the JIRA story for the helm-operator enhancement to contain this conversation.